jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -266,10 +305,32 @@ private void validateMemScope(ResourceScope memScope) throws IllegalStateExcepti
 	/* Intended for memSegmtOfPtrToLongArgFilter that converts the memory segment
 	 * of the passed-in pointer argument to long.
 	 */
-	private final long memSegmtOfPtrToLongArg(MemorySegment argValue) throws IllegalStateException {
-		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue);
+	private final long memSegmtOfPtrToLongArg(MemorySegment argValue, LinkerOptions options) throws IllegalStateException {
+		UpcallMHMetaData.validateNativeArgRetSegmentOfPtr(argValue, options);
 		addMemArgScope(argValue.scope());
-		return argValue.address();
+
+		long address = argValue.address();
+		/*[IF JAVA_SPEC_VERSION >= 22]*/
+		if (!argValue.isNative() && options.allowsHeapAccess()) {
+			HeapArgInfo info = heapArgInfo.get();
+			if (info == null) {
+				info = new HeapArgInfo(argLayoutArray.length);
+				heapArgInfo.set(info);
+			}
+			/* Store the heap argument's base object plus its non-zero offset to the first
+			 * element of the heap array as macros are no longer used in native to compute
+			 * the final address given to the foreign function.
```

Comment:
The offset doesn't necessarily refer to the first element of the array; in the case of a slice, it gives the offset to the first element included in the slice.

I don't think it is appropriate to say "macros are no longer" used; that was just an intermediate point in the journey of this pull request. Once it is merged, that will only serve to mislead future readers.

Reply:
Updated the comment to avoid confusion as discussed with Keith offline.